Libraries/LibLine/Editor.cpp

```diff
@@ -126,6 +126,36 @@ static size_t codepoint_length_in_utf8(u32 codepoint)
     return 3;
 }
 
+Editor::CodepointRange Editor::scan_for_byte_offsets(size_t start_offset, size_t end_offset, size_t scan_offset, bool reverse) const
+{
+    size_t byte_offset = 0;
+    size_t codepoint_offset = scan_offset + reverse;
+    CodepointRange offsets;
```

Comment:
`start_offset`, `end_offset`, `scan_offset`, `byte_offset`, `codepoint_offset`, `offsets`, oh my!

If you tell me what this method does, I'll think of better names for these all.

Reply:
This method converts a range described by relative byte offsets ({start,end}_offset) to a codepoint offset (`offsets`), `scan_offset` is where it should start from, and reverse is quite clear.

Reply:
OK, it's taken me some more reads of the function even after your explanation, but I think I get it now.

So the first reason why it's confusing is you call both byte offsets and codepoint indexes *offsets*. Don't, I guess?

What you do is you iterate, over a UTF-32 buffer (`m_buffer`), from `scan_offset` (which is a codepoint index, not a byte offset) either forwards or backwards, and you're counting how many bytes this would take if encoded in UTF-8, without actually encoding it in UTF-8. You're interested in the range of codepoint indexes for a range of byte offsets, `start_offset` and `end_offset`. Is that right?

I would call the method `offset_range_to_codepoint_range()`. I still don't understand why you want to scan in reverse; perhaps instead of scanning from `index_to_scan_from` down to 0 (you're always scanning to the end, don't you?), I'd scan (from 0) up until I find the codepoint index for `end_offset` and stop there. So I wouldn't need the other two arguments. Or I'm misunderstanding how this is supposed to work.

Reply:
my brain just calls relative indices offsets, I can't help it (hence why your help on naming is greatly appreciated!)

on `reverse==true`, you're right, the scan is always from 0.
but otherwise, the scan is not from the beginning or the end of the buffer.
the buffer might be N codepoints long, but the reverse scan might be initiated at codepoint offset M (M < N).

Reply:
Here's a nice diagram:
```
buffer [ 0 1 2 3 . . . A . . . B . . . M . . . N ]
                       ^       ^       ^       ^
                       |       |       |       +- end of buffer
                       |       |       +- scan offset
                       |       +- range end
                       +- range start 
```

Reply:
Yes, but instead of M→0 you could be scanning 0→M, right? And then instead of passing M explicitly, you could make the method stop as soon as it hits `end_offset`.

Reply:
Nice diagram, perhaps put that, and some of the prose, into the code as comments?

Reply:
How would that work?
I need offsets from M, not from 0

Reply:
Sure :)

Reply:
Ah, you need offsets from M! That wasn't at all clear!

OK, at this point I'd add tons of comments about all of this :smile:

Reply:
If the comments (now) included are not clear enough, please tell me so I can try clarify more~